### ISSUE TITLE
Add Visualization of Classification Results with GradCam

### ISSUE THREAD

**Issue by faizanahemad** [faizanahemad/FastNet#2]
**Proposed Features**

- Add Grid-like visualization for seeing Classification Results.
- Support for Gradcam in same visualizations.
    - add a parameter flag which enables/disables gradcam
- Visualize any model layer with Gradcam.
- Support both Loop style and Keras model.fit style models.
- Choose N samples - `examples` parameter
- Filter visualisations by:
    - visualise particular actual class only
    - visualise particular predicted class only
    - visualise particular actual class predicted as particular predicted class only
    - visualise randomly
    - misclassified-only
    - correct-only
- show histogram of prediction probabilities [Example Section:"Showing 25 Misclassified Examples"](https://github.com/faizanahemad/data-science-utils/blob/master/data_science_utils/vision/keras/SampleAPIUsage.ipynb).

**How the API looks like?**

```python
def visualize_results(model, dataset, class_names, layer=None, enable_gradcam=False,
                       actual_class=None, predicted_class=None,show_misclassified=False,show_correct=False,
                       examples=25,examples_per_row=5, image_size_multiplier=3,
                       show_actual=True, show_heatmap=False, show_superimposed=True, show_probability_histogram=False):
    ## Code here
    
```

**Intended Usage**
Once model is trained this function can be called with Model and dataset to visualize.

**Limitations**

**Online References/Documentation was referred while making this API**
- [Gradcam Examples for correct and misclassified](https://github.com/faizanahemad/data-science-utils/blob/master/data_science_utils/vision/keras/GradCam_with_misclassified_LRFinder.ipynb)
- [Gradcam examples of layers](https://github.com/faizanahemad/data-science-utils/blob/master/data_science_utils/vision/keras/Cutout_Gradcam_example.ipynb)
- [Assignment 11 Gradcam](https://github.com/faizanahemad/eva/blob/master/assignment-11/Assignment_11.ipynb)
- [Showing 25 Misclassified Examples](https://github.com/faizanahemad/data-science-utils/blob/master/data_science_utils/vision/keras/SampleAPIUsage.ipynb)

- [Code Gradcam](https://github.com/faizanahemad/data-science-utils/blob/master/data_science_utils/vision/keras/gradcam.py)
- [Code Showing 25 Misclassified Examples](https://github.com/faizanahemad/data-science-utils/blob/master/data_science_utils/vision/keras/__init__.py)

- Put these under the inspect module https://github.com/faizanahemad/FastNet/tree/master/fastnet/inspect

- While we will expose only 1 interface outside, make the inner functions modular. See the Code examples for some modularization that I had done.

**Branch Name** (The github branch in which you are working on this).

Note: Before raising a PR make sure that you have made an issue like this with tag: enhancement, ensure your code is tested and documented.

**Comment by faizanahemad** [comment]
Note: this task requires reading a lot of code

**Comment by unography** [comment]
I just had a suggestion, can we use Python 3's type annotation everywhere? That would make functions little easier to read?

something like -
```
import tensorflow as tf
from typing import List

def visualize_results(model: tf.keras.Model, class_names: List[str]):
    return
```

**Comment by faizanahemad** [comment]
Yes @unography that would be good. You can edit the original task with these annotations

**Comment by faizanahemad** [comment]
Break this into Subtasks if this seems too big to do by 1 person or in 1 week.
